sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/StreamMetrics.scala

```diff
@@ -0,0 +1,240 @@
+ */
+class StreamMetrics(sources: Set[Source], triggerClock: Clock, codahaleSourceName: String)
+  extends CodahaleSource with Logging {
+
+  import StreamMetrics._
+
+  // Trigger infos
+  private val triggerStatus = new mutable.HashMap[String, String]
+  private val sourceTriggerStatus = new mutable.HashMap[Source, mutable.HashMap[String, String]]
+
+  // Rate estimators for sources and sinks
+  private val inputRates = new mutable.HashMap[Source, RateCalculator]
+  private val processingRates = new mutable.HashMap[Source, RateCalculator]
+
+  // Number of input rows in the current trigger
+  private val numInputRows = new mutable.HashMap[Source, Long]
+  private var currentTriggerStartTimestamp: Long = -1
+  private var previousTriggerStartTimestamp: Long = -1
+  private var latency: Option[Double] = None
+
+  override val sourceName: String = codahaleSourceName
+  override val metricRegistry: MetricRegistry = new MetricRegistry
+
+  // =========== Initialization ===========
+
+  // Metric names should not have . in them, so that all the metrics of a query are identified
+  // together in Ganglia as a single metric group
+  registerGauge("inputRate-total", currentInputRate)
+  registerGauge("processingRate-total", () => currentProcessingRate)
+  registerGauge("latency", () => currentLatency().getOrElse(-1.0))
+
+  sources.foreach { s =>
+    inputRates.put(s, new RateCalculator)
+    processingRates.put(s, new RateCalculator)
+    sourceTriggerStatus.put(s, new mutable.HashMap[String, String])
+
+    registerGauge(s"inputRate-${s.toString}", () => currentSourceInputRate(s))
+    registerGauge(s"processingRate-${s.toString}", () => currentSourceProcessingRate(s))
+  }
+
+  // =========== Setter methods ===========
+
+  def reportTriggerStarted(triggerId: Long): Unit = synchronized {
+    numInputRows.clear()
+    triggerStatus.clear()
+    sourceTriggerStatus.values.foreach(_.clear())
+
+    reportTriggerStatus(TRIGGER_ID, triggerId)
+    sources.foreach(s => reportSourceTriggerStatus(s, TRIGGER_ID, triggerId))
+    reportTriggerStatus(ACTIVE, true)
+    currentTriggerStartTimestamp = triggerClock.getTimeMillis()
+    reportTriggerStatus(START_TIMESTAMP, currentTriggerStartTimestamp)
+  }
+
+  def reportTriggerStatus[T](key: String, value: T): Unit = synchronized {
+    triggerStatus.put(key, value.toString)
+  }
+
+  def reportSourceTriggerStatus[T](source: Source, key: String, value: T): Unit = synchronized {
+    sourceTriggerStatus(source).put(key, value.toString)
+  }
+
+  def reportNumInputRows(inputRows: Map[Source, Long]): Unit = synchronized {
+    numInputRows ++= inputRows
+  }
+
+  def reportTriggerFinished(): Unit = synchronized {
+    require(currentTriggerStartTimestamp >= 0)
+    val currentTriggerFinishTimestamp = triggerClock.getTimeMillis()
+    reportTriggerStatus(FINISH_TIMESTAMP, currentTriggerFinishTimestamp)
+    reportTriggerStatus(STATUS_MESSAGE, "")
+    reportTriggerStatus(ACTIVE, false)
```

Comment:
What is the meaning of `ACTIVE`?  It seems it is different than `state == ACTIVE`?